### ISSUE TITLE
Java scanner questionable handling of JAVACLASSPATH

### ISSUE THREAD

**Issue by mwichmann** [SConsProject/scons#4243]
The handling of `JAVACLASSPATH` in the Java scanner by the `_subst_libs` function seems problematic. `_subst_libs` is a direct copy of the code in `SCons/Scanner/Prog.py` (down to the comments), where the `LIBS` construction variable is modified, but `JAVACLASSPATH` is a different beast: it can be base names for class files, which is somewhat library-like, but it can also be pathnames to search for class files. The problem area is that whatever the form it's passed in, `JAVACLASSPATH` is split on whitespace, this cannot be avoided as even if a list is passed, as it recursively splits the contents.  As a result, no path with a space in it is going to work.

A potential solution is to not split if a list is passed, which would align with a lot of other SCons behavior. Not even entirely sure that splitting if it's a string is correct. For example, the unit test does something like this:

```py
    def runTest(self):
        env = DummyEnvironment(JAVASUFFIXES=['.java'],  JAVACLASSPATH=[test.workpath()])
        s = SCons.Scanner.Java.JavaScanner()
```
So it actually passes a list.  But what if you did this:
```py
        env = DummyEnvironment(JAVASUFFIXES=['.java'],  JAVACLASSPATH=test.workpath('My Java'))
```
That is, generate an absolute path whose final component has a space in it, and *not* pass that path in a list - that would fail as splitting on space there gives you a list containing an absolute path ending in `My`, and a relative path `Java`. 

Nevertheless, we could certainly decide that a string with spaces is split, but a string with spaces as a list element is not, and document that, so a way forward is at least clear.  Note that the current manpage wording is a bit murky - it could make clearer that SCons adds the search path separator before it is emitted on the command line; the user should not use that separator in setting `JAVACLASSPATH`.

Note: this was originally raised in a Discord message https://discord.com/channels/571796279483564041/571796280146133047/1030578995953152120

**Comment by mwichmann** [comment]
Additional comment as far as the manpage entry: it suggests that `JAVACLASSPATH` is always directories, but the docstring in the Java scanner code has a more expansive set of rules:
```py
    """Scan for files on the JAVACLASSPATH.

    The classpath can contain:
     - Explicit paths to JAR/Zip files
     - Wildcards (*)
     - Directories which contain classes in an unnamed package
     - Parent directories of the root package for classes in a named package

     Class path entries that are neither directories nor archives (.zip
     or JAR files) nor the asterisk (*) wildcard character are ignored.
    """
```
Should more of this be reflected in the manpage?  The manpage only talks about its use in being emitted on the `javac` command line,

**Comment by mwichmann** [comment]
Some more thought and investigation: I don't think `JAVACLASSPATH` should be space-split in the scanner at all, because such a path isn't handled by the rest of the SCons code - where the command line is built up, so it will be inconsistently interpreted.  A quick hack to the one test case that touches this to use what we might call the "splittable style":
```py
env.JavaH(target='outdir', source=[j1, j2], JAVACLASSPATH='class2 class3')
```
leads to
```console
javah -d outdir -classpath class1:class2 class3 com.Example1 com.Example2
```
Which gives an error, because `class3` doesn't end up considered part of the single option-argument to `-classpath`.  Meanwhile, issuing it in list form:
```py
env.JavaH(target='outdir', source=[j1, j2], JAVACLASSPATH=['class2', 'class3'])
```
is handled properly:
```console
javah -d outdir -classpath class1:class2:class3 com.Example1 com.Example2
```

**Comment by garfieldnate** [comment]
I've linked the commit in [Soar](SoarGroup/Soar) that led to my raising this issue.

We were joining together paths to jar files using the OS path separator, as is idiomatic for Java. The jar files already exist, and we were not expecting any dependencies to be created from the classpath (this behavior is also not documented). It's possible that with SCons 4.0.1 no Java classpath scanning was being performed, which would be why this used to work.

This is the error message we received from SCons:

    scons: *** [build\Java\Debugger\classes\edu\umich\soar\debugger\Application.class] Implicit dependency `Java\Debugger\lib\jcommon-1.0.10.jar;D:\a\Soar\Soar\Java\Debugger\lib\jfreechart-1.0.6.jar;D:\a\Soar\Soar\Java\Debugger\lib\jfreechart-1.0.6-swt.jar;D:\a\Soar\Soar\Java\Debugger\lib\swtgraphics2d.jar;D:\a\Soar\Soar\Java\Debugger\lib\swt\win64\swt.jar;D:\a\Soar\Soar\out\java\sml.jar;D:\a\Soar\Soar\out\java\soar-debugger-api.jar' not found, needed by target `build\Java\Debugger\classes\edu\umich\soar\debugger\Application.class'.

Basically, because we joined with `;` instead of a space, the entire classpath was considered one file name and added as an implicit build dependency. Then the build failed when this weird file could not be found. Even if we had joined the paths with spaces, the current SCons logic would still fail on paths that themselves already contain spaces.

**Comment by mwichmann** [comment]
I should have perhaps remembered this, but the Java-specific scanner which looks inside the classpath is new - commit bc71b14.  So not having seen this before makes sense.

Don't see any reason it couldn't spit on the search-path separators instead.  Normally, we'd expect SCons to do the assembly of the argument for the command line (by supplying a list instead of a string), since it can be platform-independent that way.

**Comment by mwichmann** [comment]
@djh82 any thoughts?

**Comment by mwichmann** [comment]
I have what I think is an improvement here - the function (renamed since it has nothing to do with libs) could look like this:
```py
def _subst_paths(env, paths) -> list:
    """Return a list of substituted path elements.

    If *paths* is a string, it is split on search-path separators - both
    the Windows and POSIX versions are tried.  Otherwise, substitution
    is done on string-valued list elements but not splitting.
    """
    if is_String(paths) -> list:
        paths = env.subst(paths)
        if SCons.Util.is_String(paths):
            if ':' in paths:
                paths = paths.split(':')
            if ';' in paths:
                paths = paths.split(';')
    else:
        paths = flatten(paths)
        paths = [env.subst(path) if is_String(path) else path for path in paths]
    return paths
```

**Comment by garfieldnate** [comment]
Although I've never seen a semicolon or a colon in a path anywhere, would it not be safer to only split on the os-specific path separator?

**Comment by mwichmann** [comment]
Could do that - I guess if you use the wrong one it's going to break when it's passed to Java anyway, so we could just assume the right flavor for the system has been passed?

**Comment by djh82** [comment]
I didn't expect folk to be constructing the classpath string themselves - might as well let scons do it for you so you can be platform agnostic. Constructing a very specifically formatted string for a tool is very un-scons like - normally the tools handle all that for you (though the java tooling isn't as good as some of the others). And even if those files aren't built by scons, you probably do want a rebuild if they change?


I might split on `os.pathsep` though, since in the windows case you can end up with drive letters being split by the colon case unexpectedly, which might have unintended consequences?

**Comment by garfieldnate** [comment]
>Although I've never seen a semicolon or a colon in a path anywhere

🤦  yes of course, drive letters on Windows! We should not split on those.

**Comment by mwichmann** [comment]
I changed it for the PR to just use `os.pathsep` (see #4244)

**Comment by mwichmann** [comment]
Yes, if someone passed such. My read (and it's only a deduction, I'm not in that author's head :) ) is that `LIBS="glib gtk-2.0 pango"` was a plausible use case, and so here split-on-space made more sense.  I haven't even checked if the rest of SCons would handle such a `LIBS` specification.

**Comment by djh82** [comment]
Yes, I guess that's OK because LIBS and LIBPATH are separate env variables, where LIBS should just be names to find on the LIBPATHs. Whereas JAVACLASSPATH combines both into one and therefore suffers with this issue.
